scalyr_agent/util.py

```diff
@@ -1315,25 +1353,60 @@ def _is_running(self):
     # to become available.
     BUSY_LOOP_POLL_INTERVAL = .03
 
-    def _sleep_for_busy_loop(self, deadline, description):
+    def _sleep_for_busy_loop(self, deadline, last_loop_time, description):
         """Sleeps for a small unit of time as part of a busy wait loop.
 
         This method will return if either the small unit of time has exceeded, the overall deadline has been exceeded,
         or if the `stop` method of this thread has been invoked.
 
-        @param deadline: The walltime that this operation should time out.  This method will sleep for the small of
-            BUSY_LOOP_POLL_INTERVAL or the difference between now and walltime.
+        @param deadline: The walltime that this operation should time out.  This method will sleep until the smaller of
+            last_loop_time + BUSY_LOOP_POLL_INTERVAL or deadline.
+        @param last_loop_time: The time the last loop through the busy wait loop began.  This is used to calculate the
+            deadline of the busy sleep.  Note, it is also important for catching advances in the fake clock when
+            in test mode.
         @param description: A description of why we waiting to be used in error output.
 
         @type deadline: float
+        @type last_loop_time: float
         @type description: str
         """
-        timeout = deadline - self.__time()
-        if timeout < 0:
+        current_time = self.__time()
+        poll_deadline = RedirectorClient.BUSY_LOOP_POLL_INTERVAL + last_loop_time
+
+        if deadline - current_time < 0:
             raise RedirectorError('Deadline exceeded while waiting for %s' % description)
-        elif timeout > RedirectorClient.BUSY_LOOP_POLL_INTERVAL:
-            timeout = RedirectorClient.BUSY_LOOP_POLL_INTERVAL
-        self._run_state.sleep_but_awaken_if_stopped(timeout)
+        elif deadline > poll_deadline:
+            deadline = poll_deadline
+
+        if self.__fake_clock is not None:
+            self._run_state.sleep_but_awaken_if_stopped(deadline - current_time)
+        else:
+            self.__simulate_busy_loop(deadline)
```

Comment:
I might be reading this wrong but it seems to me this line only executes if `self.__fake_clock` is `None`, in which case line 1406 always raises exception?

Reply:
The greater question I have is why are we simulating a busy loop if the fake clock is `None`?  Shouldn't w be sleeping for a real duration of time?